### PR TITLE
Chapter: Related Work / Section: other-approaches

### PR DESCRIPTION
**Pull request by ysndr**
- [x] Extensions
- [x] LSIF
- [x] DAP

### CODE REVIEW

**Review by yannham — COMMENTED** [review]
I found this part harder to read. I'm sorry, I don't have very precise suggestions to make it better, but I feel like it could be made much shorter and straight to the point while retaining the important information. Several sentences are long or overly general/abstract. Maybe adding a few examples or illustrations (I don't necessarily mean graphics, but illustration of ideas or concepts) could do some good.

File: chapter/related-work.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+The LSP defines a large range of commands and capabilities which is continuously being extended by the maintainers of the protocol.
+Yet, occasionally server developers find themselves in need of functionality not yet present in the protocol.
+For example the LSP does not provide commands to exchange binary data such as files.
+In [@sec:language-independence] the CodeCompass Language Server was introduced.
+A stern feature of this server is the ability to generate and show diagrams in SVG format.
+However, the LSP does not define the concept of *requesting diagrams*.
+In particular Mészáros et al. describe different shortcomings of the LSP :
+
+1. The "LSP doesn’t have a feature to place a context menu at an arbitrary spot in the document"
```
[the remark this answers is not in the document]
Comment (ysndr) [reply chapter/related-work.md:49]: I think they are rendered quite weirdly in latex but I can try.
Otherwise, I thought the same

File: chapter/related-work.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+### Language Server Index Format
+
+The Language Server Index Format, or short LSIF, is an augmentation of the LSP.
+Since the LSP requires a language server to actively analyze files and answer requests, its use is typically constrained to local installations.
+As pointed out in the introducing blog post [@lsif-blog-post], several use cases exist where the LSP approach fails due to resource limits.
+The article explicitly names web based platforms such as GitHub[@github] or Sourcegraph[@sourcegraph].
+
+The LSIF aims to provide the features of the LSP without the need of actively running a language server.
+Instead, "language servers or other programming tools to emit their knowledge about a code workspace" as a LSIF compliant JSON report.
```
[the remark this answers is not in the document]
Comment (ysndr) [reply chapter/related-work.md:83]: I think the "to" in the quote is too much and would make the sentence flow much better i strike it
```suggestion
Instead, "language servers or other programming tools [..] emit their knowledge about a code workspace" as a LSIF compliant JSON report.
```

File: chapter/related-work.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+The specification [@lsif-spec] defines four principal goals for the LSIF:
+
+- The format should not imply the use of a certain persistence technology.
+- The data defined should be modeled as closely as possible to the Language Server Protocol to make it possible to serve the data through the LSP without further transformation.
+- The data stored is result data usually returned from an LSP request. 
+- The output format will be based on JSON as with the LSP.
+
+The format specifies a graph structure that comprises that links ranges of source code to language analysis results that are based on the data types defined by the LSP.
```
[the remark this answers is not in the document]
Comment (ysndr) [reply chapter/related-work.md:92]: yes
```suggestion
The format specifies a graph structure that links ranges of source code to language analysis results.
Being based on the LSP the graph layout tries to resemble LSP types.
```

File: chapter/related-work.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ysndr) [reply chapter/related-work.md:49]:
```suggestion
1. > "LSP doesn’t have a feature to place a context menu at an arbitrary spot in the document"
```

[the remark this answers is not in the document]
Comment (ysndr) [reply chapter/related-work.md]: Lets find out

File: chapter/related-work.md [hunk 4]
```diff
@@ -36,10 +36,94 @@ Additionally, this part aims to recognize alternative approaches to the LSP, in
...
+1. > "LSP doesn’t have a feature to place a context menu at an arbitrary spot in the document"
+   Context menu entries are implemented by clients based on the agreed upon capabilities of the server.
+   Undefined capabilities cannot be added to the context menu.
+
+   In the case of CodeCompass the developers made up for this by using the code completion feature as an alternative dynamic context menu.
+
+2. "LSP does not support displaying pictures (diagrams)".
+   CodeCompass generates diagrams for selected code.
+   Yet, there is no image transfer included with the LSP.
+   Since the LSP is based on JSON-RPC messages, the authors' solution was to define a new command, specifically designed to tackle this non-standard use case.
```
Comment (ysndr) [inline chapter/related-work.md:58]:
```suggestion
1. > "LSP doesn’t have a feature to place a context menu at an arbitrary spot in the document"

   Context menu entries are implemented by clients based on the agreed upon capabilities of the server.
   Undefined capabilities cannot be added to the context menu.

   In the case of CodeCompass the developers made up for this by using the code completion feature as an alternative dynamic context menu.

2. > "LSP does not support displaying pictures (diagrams)".

   CodeCompass generates diagrams for selected code.
   Yet, there is no image transfer included with the LSP.
   Since the LSP is based on JSON-RPC messages, the authors' solution was to define a new command, specifically designed to tackle this non-standard use case.
```

File: chapter/related-work.md [hunk 5]
```diff
@@ -36,10 +36,96 @@ Additionally, this part aims to recognize alternative approaches to the LSP, in
...
+
+### Language Server Index Format
+
+The Language Server Index Format, or short LSIF, is an augmentation of the LSP.
+Since the LSP requires a language server to actively analyze files and answer requests, its use is typically constrained to local installations.
+As pointed out in the introducing blog post [@lsif-blog-post], several use cases exist where the LSP approach fails due to resource limits.
+The article explicitly names web based platforms such as GitHub[@github] or Sourcegraph[@sourcegraph].
+
+The LSIF aims to provide the features of the LSP without the need of actively running a language server.
+Instead, "language servers or other programming tools to emit their knowledge about a code workspace" as a LSIF compliant JSON report.
```
Comment (ysndr) [inline chapter/related-work.md:85]:
```suggestion
Most commonly, web sites mainly present code instead of offering rich editor experiences.
Nevertheless, platforms such as GitHub[@github] or Sourcegraph[@sourcegraph] could greatly benefit from LSP like features but are practically constrained by the need of a LSP Server
Since the presented code is static related metadata could be preprocessed.
The LSIF (Language Server index Format) specifies a schema for output of such a preprocessing.
Clients in turn, can provide code intelligence using statically provided, LSIF formatted reports, instead of needing to communicate with an active language server.
```

How about rewording like so?

File: chapter/related-work.md [hunk 6]
```diff
@@ -36,10 +36,96 @@ Additionally, this part aims to recognize alternative approaches to the LSP, in
...
+
+The LSIF specification [@lsif-spec] defines four principal goals:
+
+- The format should not imply the use of a certain persistence technology.
+- The data defined should be modeled as closely as possible to the Language Server Protocol to make it possible to serve the data through the LSP without further transformation.
+- The data stored is result data usually returned from an LSP request. 
+- The output format will be based on JSON as with the LSP.
```
Comment (ysndr) [inline chapter/related-work.md:92]: @yannham i don't know if this is actually that relevant. Although it is important fot he lsif it does not contribute to understanding the benefits of the format

File: chapter/related-work.md (hunk 5 above)
Comment (yannham) [inline chapter/related-work.md:85]: Yep, better. Here is another suggestion, to clarify a few things (e.g. static code -> you can't edit the code online), feel free to use/modify/merge with your suggestion/drop.

```suggestion
Nowadays, code hosting platforms are an integral part of the developer toolset (GitHub[@github], Sourcegraph[@sourcegraph], GitLab, Sourceforge, etc.). Those platforms commonly displays code simply as text, highlighted at best.
LSP-like features would make for a great improvement for code navigation and code reading online, but analysis would be very wasteful if it needed to be done each time a visitor load a chunk of code. Since the hosted code is most often not editable and precisely versioned, code analysis could be performed ahead of time, once and for all for each version.
The LSIF (Language Server index Format) specifies a schema for the output of such ahead of time code analysis.
Clients can then provide efficient code intelligence using LSIF data provided in one shot by the server.
```

File: chapter/related-work.md (hunk 6 above)
Comment (yannham) [reply chapter/related-work.md:92]: I would say it's your call. It provides interesting information. But if all you want to say about LSIF is to explains roughly what it is and what problems it tries to solve, it is also reasonable to get rid of this part, as it is quite specific.

File: chapter/related-work.md [hunk 7]
```diff
@@ -36,10 +36,97 @@ Additionally, this part aims to recognize alternative approaches to the LSP, in
...
+Their approach is based on a generic intermediate representation of graphs which can be generated by language servers and turned into a graphical representation by the client implementation.
+
+Similarly, in [@Specification-Language-Server-Protocol,@decoupling-core-analysis-support] the authors describe a method to develop language agnostic LSP extensions.
+In their work they defined a language server protocol for specification languages (SLSP) which builds on top of the existing LSP, but adds several additional commands.
+The commands are grouped by their use case in the domain of specification languages and handled by separate modules of the client extension implementing support for the SLSP.
+Following the LSP example and defining commands language agnostic instead of tied to a specific language, allows to maintain the initial purpose of the LSP.
+Since the extensions can be incorporated by specific implementations of language servers in the same domain, a single client implementation serves multiple languages.
+The authors point out that while their approach specializes in specification languages, the idea can be transferred to other areas.
```
Comment (ysndr) [inline chapter/related-work.md:75]:
```suggestion
Similarly, in [@Specification-Language-Server-Protocol,@decoupling-core-analysis-support] the authors describe a method to develop language agnostic LSP extensions.
In their work they defined a language server protocol for specification languages (SLSP) which builds on top of the existing LSP.
The SLSP defines several extensions that each group the functionality of specific domains.
However, unlike other LSP extensions that are added to facilitate functions of a specific server, SLSP is language agnostic.
In effect, the protocol extensions presented by Rask et. al. are reusable across different specification languages, allowing clients to implements a single frontend.
Given their successes with the presented work, the authors encourage to build abstract, sharable extensions over language specific ones if possible.
```

File: chapter/related-work.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+Their approach is based on a generic intermediate representation of graphs which can be generated by language servers and turned into a graphical representation by the client implementation.
+
+Similarly, in [@Specification-Language-Server-Protocol,@decoupling-core-analysis-support] the authors describe a method to develop language agnostic LSP extensions.
+In their work they defined a language server protocol for specification languages (SLSP) which builds on top of the existing LSP, but adds several additional commands.
+The commands are grouped by their use case in the domain of specification languages and handles by separate modules of the client extension implementing support for the SLSP.
+Following the LSP example and defining commands language agnostic instead of tied to a specific language, allows to maintain the initial purpose of the LSP.
+Since the extensions can be incorporated by specific implementations of language servers in the same domain, a single client implementation serves multiple languages.
+The authors point out that while their approach specializes in specification languages, the idea can be transferred to other areas.
```
[the remark this answers is not in the document]
Comment (ysndr) [reply chapter/related-work.md:75]: https://github.com/ysndr/MSc-Thesis/pull/19/files#r841078857

File: chapter/related-work.md [hunk 9]
```diff
@@ -36,10 +36,97 @@ Additionally, this part aims to recognize alternative approaches to the LSP, in
...
+Since the hosted code is most often static and precisely versioned, code analysis could be performed ahead of time, for all files of each version.
+The LSIF (Language Server index Format) specifies a schema for the output of such ahead of time code analysis.
+Clients can then provide efficient code intelligence using LSIF data provided in one shot by the server.
+
+The LSIF specification [@lsif-spec] defines four principal goals:
+
+- The format should not imply the use of a certain persistence technology.
+- The data defined should be modeled as closely as possible to the Language Server Protocol to make it possible to serve the data through the LSP without further transformation.
+- The data stored is result data usually returned from an LSP request. 
+- The output format will be based on JSON as with the LSP.
```
Comment (ysndr) [inline chapter/related-work.md:92]:
```suggestion

```
